# 01. Importing libraries

In [1]:
# Import libraries
import warnings
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [2]:
# Define csv path
path = r'C:\Users\lucav\Desktop\Instacart Basket Analysis'

In [3]:
# Import data from 'orders_checked.csv'
warnings.simplefilter(action='ignore', category=FutureWarning)
df_ords = pd.read_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'), index_col = 0)

In [4]:
# Import data from 'products_checked.csv'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = 0)

In [5]:
# Import data from 'orders_products_combined.pkl'
orders_products_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

# 03. Task

## I. Check the shape of the imported dataframe

In [6]:
orders_products_combined.shape

(32434489, 10)

The dimensions are the same as those of the exported file.

## II. Combine 'orders_products_combined' and 'df_prods'

In [7]:
# Return the first 5 rows of 'orders_products_combined'
orders_products_combined.head()

order_id  user_id  order_number  day_of_week  hour_of_day  \
0   2539329        1             1            2            8   
1   2539329        1             1            2            8   
2   2539329        1             1            2            8   
3   2539329        1             1            2            8   
4   2539329        1             1            2            8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered  
0          0  
1          0  
2          0  
3          0  
4          0

In [8]:
# Return the first 5 rows of 'df_prods'
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [9]:
# Combine 'orders_products_combined' and 'df_prods' on 'product_id'
df_orders_products_combined_products = orders_products_combined.merge(df_prods, on = 'product_id', indicator = True)

In [10]:
# Return the first 5 rows of 'df_orders_products_combined_products'
df_orders_products_combined_products.head()

order_id  user_id  order_number  day_of_week  hour_of_day  \
0   2539329        1             1            2            8   
1   2398795        1             2            3            7   
2    473747        1             3            3           12   
3   2254736        1             4            4            7   
4    431534        1             5            4           15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  
0          0         Soda        77              7     9.0   both  
1          1         Soda        77              7     9.0   both  
2          1         Soda        77              7     9.0   both  
3          1         Soda        77              7     9.0   both  
4          1         Soda        77              7     9.0   both

In [11]:
# Check for full matches
df_orders_products_combined_products['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

In [12]:
# Remove '_merge' column
df_orders_products_combined_products_final = df_orders_products_combined_products.drop(columns = ['_merge'])

In [13]:
# Return the first 5 rows of 'df_orders_products_combined_products_final'
df_orders_products_combined_products_final.head()

order_id  user_id  order_number  day_of_week  hour_of_day  \
0   2539329        1             1            2            8   
1   2398795        1             2            3            7   
2    473747        1             3            3           12   
3   2254736        1             4            4            7   
4    431534        1             5            4           15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered product_name  aisle_id  department_id  prices  
0          0         Soda        77              7     9.0  
1          1         Soda        77              7     9.0  
2          1         Soda        77              7     9.0  
3          1         Soda        77              7     9.0  
4          1         Soda        77              7     9.0

In [14]:
# Return the dimensions of 'df_orders_products_combined_products_final'
df_orders_products_combined_products_final.shape

(32404859, 14)

Note that the 'orders_products_combined' dataframe initially had 32434489 rows, but the result of the merge with the df_prods dataframe returns only 32404859 rows. Thus, the two dataframes are not a full match as some observations are only available on either of the two.

## III. Export 'df_orders_products_combined_products_final' as 'orders_products_merged.pkl'

In [15]:
df_orders_products_combined_products_final.to_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_merged.pkl'))